Source/DafnyCore/Resolver/PreType/PreTypeConstraints.cs

```diff
@@ -0,0 +1,546 @@
+//-----------------------------------------------------------------------------
+//
+// Copyright by the contributors to the Dafny Project
+// SPDX-License-Identifier: MIT
+//
+//-----------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Diagnostics.Contracts;
+
+namespace Microsoft.Dafny {
+  /// <summary>
```

Comment:
I think this comment should try to give me enough knowledge, possibly through links to other documents, to be able to easily read and understand this class, which it doesn't do for me yet.

Could you explain why we have a constraint solving and a decision phase?

Reply:
Excellent question. In this use of the words, "constraint solving" just sorts out things that follow from the given constraints, whereas "decisions" makes choices between two or more ways to proceed. For example, if `a` is known to be an integer, then the constraint imposed by a program expression `a == b` says that `b` has to be an integer as well. In contrast, an assignment like `c := 10;` has a choice in what type to give to the literal `10` and `c`, because the type could be `int`, `bv32`, `ORDINAL`, or many other things. Once a decision (that is, a choice) has been made, the rest of the type inference must abide by that decision, so it's good to precede decisions with constraint solving.

The reason there is sometimes constraint solving that is _not_ followed by making decisions (in order words, what the `PartiallySolve...` method does) is that there is a small number of Dafny expressions where the type inference wants to look at the consequences of the constraints laid down so far, but without making any decisions. The prime example of this is member lookup, like in the expression `o.M`. The old type inference in Dafny always insisted on having the ("head" part of the) type of `o` when encountering `o.M`; if it couldn't determine this type, then it would immediately give an error. To determine the type of `o` in this case, one typically needs constraint solving. If that's sufficient, then it's best not to make any hasty decisions at that time.

In the new type inference, there is a general facility of "guarded constraints", which are constraints waiting to happen. Using this facility, I have changed many of the old ways of urgently needing a type. Essentially, this means that the resolver can continue with other expressions and then come back to a guarded constraint, at a time when, hopefully, there is enough inferred information to evaluate the "guard". If we would change to using such guarded constraints everywhere in the type inference, then (to answer your question) we would not need a `PartiallySolve...` method. But I haven't changed the way that member expressions like `o.M` are handled (or the way that the source expression in a `match` is handled). It would be good to change type inference to use the guarded constraints for those, too, but I wanted to see that everything else worked before embarking on those large changes. Thus, for now, we still need `PartiallySolve...`. But in the future, I hope we can improve both the code and the inference itself by changing those places to use guarded constraints, too.

Here is an example program that we don't accept today (neither with the old type inference nor with `/typeSystemRefresh:1` in PR now), but that we could accept in the future if we changing member lookup to use a guarded constraint:

``` dafny
class Cell { var data: int }

method M() {
  var c := null;
  for i := 0 to 10 {
    if c != null {
      c.data := c.data + 1; // this currently gives an error, saying it doesn't know what .data is
    }
    if i == 7 {
      c := new Cell; // this line, which comes after c.data above, is what's needed to infer the type of c
    }
  }
}
```